File: dynamic_programming/knapsack/322_coin_change.cc
#include <climits>
#include <vector>

// 2024/01/12: 复习了背包问题, 且调试了25min完成.
// 完全背包问题
int CoinChange(std::vector<int>& coins, int amount) {
  // dp[i][j]: 可以使用第i个coin, 总价值正好为j.
  int m = coins.size();
  std::vector<std::vector<int>> dp(m + 1,
                                   std::vector<int>(amount + 1, INT_MAX));
  for (int i = 1; i <= m; ++i) {
    for (int j = 0; j <= amount; ++j) {
      if (j == 0) {
        dp[i][j] = 0; // 粗心! = 而非 == 
        continue;
      }
      if (j >= coins[i - 1] && dp[i][j - coins[i - 1]] != INT_MAX) {
        dp[i][j] = std::min(dp[i][j], dp[i][j - coins[i - 1]] + 1);
      }
      dp[i][j] = std::min(dp[i][j], dp[i - 1][j]); // 粗心! 需要取小.
    }
  }
  return dp[m][amount] == INT_MAX ? -1 : dp[m][amount];
}

int main(int argc, char* argv[]) {
  std::vector<int> coins = {1,2,5};
  CoinChange(coins, 11);
  return 0;
}

File: dynamic_programming/knapsack/322_coin_change.py
from typing import List

# 2026/06/07: 复习了背包问题之后自己写出来.
def coin_change(coins: List[int], amount: int) -> int:
    # dp[i][j] 表示只用前 i 类钱币且价值正好为 j 所需最小数量.
    dp = [[float("inf")] * (amount + 1) for _ in range(len(coins) + 1)]

    for i in range(1, len(coins) + 1):
        dp[i][0] = 0

    for i in range(1, len(coins) + 1):
        coin = coins[i - 1]
        for j in range(1, amount + 1):
            if j >= coin:
                dp[i][j] = min(dp[i - 1][j], dp[i][j - coin] + 1)
            else:
                dp[i][j] = dp[i - 1][j]
    return dp[-1][-1] if dp[-1][-1] != float("inf") else -1

coins = [1, 2, 5]
amount = 11

coins = [2]
amount = 3
print(coin_change(coins, amount))